### ISSUE TITLE
todo! and unimplemented! does not work with impl Trait return types

### ISSUE THREAD

**Issue by neckbosov** [rust-lang/rust#69882]
This code
```
trait SomeTrait {
    fn some_func();
}

fn todo_impl_trait() -> impl SomeTrait { todo!() }
```
does not compile because 
`the trait `SomeTrait` is not implemented for ()`
But such code

```
trait SomeTrait {
    fn some_func();
}

fn todo_impl_trait<T: SomeTrait>() -> T { todo!() }
```
compiles correctly. Can this problem be resolved to use both todo!() and impl Trait?

**Comment by jonas-schievink** [comment]
This is expected behavior since `todo!()` is not constrained to any type, so it is inferred to `()`. In the second case it is constrained to `T`, which is provided by the caller.

**Comment by robinmoussu** [comment]
Given that 4 bugs have been opened in such a short period of time for the same reason, I don't think that a wontfix, works as intented is the right approach.

At minimum an alternative should be provided in the help message from the compiler, but I personnaly think that `todo!()`, `unimplemented!()` and the like should typecheck to `!` (never), that itsel should typecheck to any `impl Trait`.

**Comment by jyn514** [comment]
> todo!(), unimplemented!() and the like should typecheck to ! (never), that itsel should typecheck to any impl Trait.

https://github.com/rust-lang/rust/blob/567ad7455d5f25f6b38d2fded1cb621e0c34a48b/src/librustc_interface/util.rs#L587-L594

**Comment by jonas-schievink** [comment]
Reopening this to have a canonical issue to direct people to. This is still working as expected for the moment though.

**Comment by jonas-schievink** [comment]
Actually, let's classify this as a diagnostics enhancement

**Comment by jeffs** [comment]
Is there any suggested work-around for this?  It sure seems like the bottom type ought (by definition) to type-check as anything, but as a practical matter, is there any placeholder at all that can be used as a return value from an unimplemented function whose signature demand an impl trait?

**Comment by jyn514** [comment]
@jeffs in the general case, no, because there could be no type that implements the trait: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=beb24c95bf6be2a8b48f6381c2b7ad6d

**Comment by birkenfeld** [comment]
> @jeffs in the general case, no, because there could be _no_ type that implements the trait: [play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=beb24c95bf6be2a8b48f6381c2b7ad6d](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=beb24c95bf6be2a8b48f6381c2b7ad6d)

I don't get what the example shows; you can add `impl Trait for u8` or something, and still get the same error.

**Comment by ede1998** [comment]
> I don't get what the example shows; you can add impl Trait for u8 or something, and still get the same error.
It shows that it's impossible to return a placeholder type because there might not be a type that impls the trait.

**Comment by tavianator** [comment]
`rustc` could make up a type that implements the trait with `unreachable!()` as the body for every `fn`.

**Comment by ede1998** [comment]
No that wouldn't be enough. The trait might also have associated types or constants, which would have to be implemented as well. 
Also the trait could be unsafe, so implementing it with unreachable might introduced undefined behavior.

**Comment by DCNick3** [comment]
Error message could definitely be improved. Currently this code:

```rust
trait Trait {}

fn magic() -> impl Trait {
    panic!()
}
```

Results in 

```
error[E0277]): the trait bound `(): Trait` is not satisfied
 --> src/lib.rs:3:15  
  |
3 | fn magic() -> impl Trait {
  |               ^^^^^^^^^^ the trait `Trait` is not implemented for `()`
```

Which is really confusing.

Using `Vec<impl Trait>` gives a message that is much better (at least saying that it has anything to do with the `!` type and the return value):

```
error[E0720]: cannot resolve opaque type
 --> src/lib.rs:3:19
  |
3 | fn magic() -> Vec<impl Trait> {
  |                   ^^^^^^^^^^ cannot resolve opaque type
4 |     panic!()
  |     -------- this returned value is of `!` type
  |
  = help: this error will resolve once the item's body returns a concrete type
```

Why is it not shown in the first case?

**Comment by oli-obk** [comment]
I want to fix this. Unfortunately if you had an `impl Trait for () {}` in your first snippet, that would compile right now and error if my fix would get accepted.

I could try to just fix the diagnostic, but it feels like just piling on hacks (the fact that this works if `()` implements the trait is already a huge hack imo)

**Comment by oli-obk** [comment]
> No that wouldn't be enough. The trait might also have associated types or constants, which would have to be implemented as well. Also the trait could be unsafe, so implementing it with unreachable might introduced undefined behavior.

A concrete example that would be unsound or at least ICE the compiler is https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ea0e808536f4373dc7be9f002159fa1d

If we just didn't register any hidden type, the `make_me` call cannot be compiled successfully.

If we used `!` fallback, we would get an error because `!` does not implement the trait. Of course we could say we generate the trait and panic in all methods, but what happens to associated types and constants? Do we invent random values?

This gets worse when we'll have type alias impl trait, as then you can directly name the opaque type and access associated items directly without jumping through hoops like in my playground.

We could special case `todo` and `unimplemented` to just pick some type that implements the trait, but that seems like a real footgun and we should totally lint those cases imo

**Comment by drdozer** [comment]
I just hit this. It makes it impossible to stub out anything that returns `impl T`. So even if you don't want to write it now, you have to, because ... reasons. Is there any way this could be partially supported for things where it won't break the world? So the compiler would have the option to say "can't do this because trait Foo has evil construct xxx"?

**Comment by oli-obk** [comment]
If you know a type that satisfies the trait bounds you can use `todo!() as TheType`. Not as convenient, but it works.

I mean... in theory we could just pick a random type automatically that satisfies the trait, but then someone would start relying on this and we have enough problems with things like the one-impl-rule which are a similar situation

**Comment by oli-obk** [comment]
Idea: we could make the error have a suggestion that suggests casting the `todo!()` to a random type that satisfies the trait (if one exists)

This way we don't create more language corner cases and allow you to fix all those todos with `cargo check --fix`

**Comment by hanna-kruppe** [comment]
Currently such a cast [gets flagged as unreachable](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c6e51352f055b2b60baaf854543fadd) (technically correct) so either that lint needs to get tweaked or the suggestion also needs to include a well-placed `expect(unreachable_code)`.

**Comment by drdozer** [comment]
> If you know a type that satisfies the trait bounds you can use `todo!() as TheType`.

Yeah, I get this, but during the initial sketching, you don't have any implementations. In an API-first design, there's nothing but the trait until the API has been sketched. So needing to stub the whole thing out with noop impls to stub any of it out is a massive pain.

Perhaps this is a "wrong problem" problem, and maybe instead there's a concept of "an unimplemented type" that's a bit like unimplemented code? Like a middle ground between "the type exists and is FOO" and "there is no type FOO", that's "There is no type FOO yet, but if there was, ...". And at some point in the compilation chain, the compilation would bail with "unimplemented types", but at least you'd be able to typecheck through them? This is probably a horrible idea for various reasons that I've not thought through.

Sorry not to be more help.

**Comment by dan-da** [comment]
I just hit this issue also.  In my case my trait has a blocking run() method and a run_async() method.  I wanted to provide an unimplemented default for each, so that trait implementor's ony need to impl appropriate run method, blocking or async.

unimplemented!() works just fine in run(), but won't compile in run_async().   This forces the trait implementor to define a run_async() even if it won't be used.  That works, but is ugly.

I illustrate the above only to describe a real use-case where this limitation is problematic.

**Comment by musjj** [comment]
@oli-obk 

> I mean... in theory we could just pick a random type automatically that satisfies the trait

Why not just return an opaque `impl T` type?
